tests/test_tga.py

```diff
@@ -171,6 +179,203 @@ def test_prefix_tokens(self):
+                    ('__start__', 0.0, 1.0),
+                    ('5', -2.5510462364763953e-05, 0.0),
+                    ('__end__', -1.1920922133867862e-06, 0.0),
+                ],
+                'extra_args': ['--beam-size', '3'],
+            },
+            'greedy': {
+                'text_token_info': [
+                    ('__start__', 0.0, 1.0),
+                    ('5', -2.5510462364763953e-05, 0.0),
+                    ('__end__', -1.1920922133867862e-06, 0.0),
+                ],
+                'extra_args': [],
+            },
+            # sampling based token selection will produce non-deterministic output, so we can't do data regression
+            'topk': {'extra_args': ['--topk', '2']},
+            'topk_multiple_beams': {'extra_args': ['--topk', '2', '--beam-size', '5']},
+            # sampling based token selection will produce non-deterministic output, so we can't do data regression
+            'nucleus': {'extra_args': ['--topp', '0.3']},
+            'nucleus_multiple_beams': {
+                'extra_args': ['--topp', '0.3', '--beam-size', '5']
+            },
+            # sampling based token selection will produce non-deterministic output, so we can't do data regression
+            'delayedbeam': {'extra_args': ['--topk', '2', '--beam-delay', '2']},
+        }
+
+        for inference_type in inference_types:
+            args = [
+                '--model-file',
+                'zoo:unittest/transformer_generator2/model',
+                '--inference',
+                inference_type,
+                '--truncate',
+                '1024',
+                '-v',
+            ] + gold_data[inference_type]['extra_args']
+
+            pp = ParlaiParser(True, True)
+            agent = create_agent(pp.parse_args(args), True)
+            obs = {'text': '5', 'episode_done': False}
+            agent.observe(obs)
+            act = agent.act()
+
+            if 'text_token_info' in gold_data[inference_type]:
+                for i, tok_data in enumerate(act['text_token_info']):
+                    assert (
+                        gold_data[inference_type]['text_token_info'][i][0]
+                        == tok_data[0]
+                    ), f"failed token prediction for inference type {inference_type} at token {gold_data[inference_type]['text_token_info'][i][0]}"
+                    assert math.isclose(
+                        gold_data[inference_type]['text_token_info'][i][1], tok_data[1]
+                    ), f"failed token probability prediction for inference type {inference_type} at token {gold_data[inference_type]['text_token_info'][i][0]}"
+                    assert math.isclose(
+                        gold_data[inference_type]['text_token_info'][i][2], tok_data[2]
+                    ), f"failed token rank prediction for inference type {inference_type} at token {gold_data[inference_type]['text_token_info'][i][0]}"
+
+    def test_tree_search(self):
```

Comment:
> i'll also add one last request that you add a test that would capture some of the bugs you ran into here. For example, if you initialize the BeamSearch object and pass a set of (fake) logprobs to select_paths, can you ensure that it returns you the correct token scores? (& similarly for the other tree search objects)

Here is a test that should capture this for all the different decoding schemes w/ and w/o beams.